From aaaba19a695803b108bf17a054927054f840ba7e Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Fri, 21 Sep 2018 13:26:03 +0200
Subject: [PATCH 099/137] ipipe: timer: do not interpose on undefined handlers

There is no point in interposing on clock chip handlers for which
there was no support originally. In some cases (oneshot_stopped), we
may even get a kernel fault, jumping to a NULL address.

Interpose on non-NULL original handlers only.
---
 kernel/ipipe/timer.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/kernel/ipipe/timer.c b/kernel/ipipe/timer.c
index 5696c28..5ba94bf 100644
--- a/kernel/ipipe/timer.c
+++ b/kernel/ipipe/timer.c
@@ -435,10 +435,14 @@ static void grab_timer(void *arg)
 		evtdev->mult = 1;
 		evtdev->shift = 0;
 		evtdev->max_delta_ns = UINT_MAX;
-		evtdev->set_state_periodic = do_set_periodic;
-		evtdev->set_state_oneshot = do_set_oneshot;
-		evtdev->set_state_oneshot_stopped = do_set_oneshot;
-		evtdev->set_state_shutdown = do_set_shutdown;
+		if (timer->orig_set_state_periodic)
+			evtdev->set_state_periodic = do_set_periodic;
+		if (timer->orig_set_state_oneshot)
+			evtdev->set_state_oneshot = do_set_oneshot;
+		if (timer->orig_set_state_oneshot_stopped)
+			evtdev->set_state_oneshot_stopped = do_set_oneshot;
+		if (timer->orig_set_state_shutdown)
+			evtdev->set_state_shutdown = do_set_shutdown;
 		evtdev->set_next_event = data->emutick;
 		evtdev->ipipe_stolen = 1;
 	}
-- 
1.9.1

